compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1037,33 +1040,82 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return;
         }
 
-        if let hir::FnRetTy::Return(ty) = fn_decl.output {
-            let ty = self.astconv().ast_ty_to_ty(ty);
-            let bound_vars = self.tcx.late_bound_vars(fn_id);
-            let ty = self
-                .tcx
-                .instantiate_bound_regions_with_erased(Binder::bind_with_vars(ty, bound_vars));
-            let ty = match self.tcx.asyncness(fn_id.owner) {
-                ty::Asyncness::Yes => self.get_impl_future_output_ty(ty).unwrap_or_else(|| {
-                    span_bug!(fn_decl.output.span(), "failed to get output type of async function")
-                }),
-                ty::Asyncness::No => ty,
-            };
-            let ty = self.normalize(expr.span, ty);
-            if self.can_coerce(found, ty) {
-                if let Some(node) = self.tcx.opt_hir_node(fn_id)
-                    && let Some(owner_node) = node.as_owner()
-                    && let Some(span) = expr.span.find_ancestor_inside(owner_node.span())
+        let in_closure = matches!(
+            self.tcx
+                .hir()
+                .parent_iter(id)
+                .filter(|(_, node)| {
+                    matches!(
+                        node,
+                        Node::Expr(Expr { kind: ExprKind::Closure(..), .. })
+                            | Node::Item(_)
+                            | Node::TraitItem(_)
+                            | Node::ImplItem(_)
+                    )
+                })
+                .next(),
+            Some((_, Node::Expr(Expr { kind: ExprKind::Closure(..), .. })))
+        );
+
+        let can_return = match fn_decl.output {
+            hir::FnRetTy::Return(ty) => {
+                let ty = self.astconv().ast_ty_to_ty(ty);
+                let bound_vars = self.tcx.late_bound_vars(fn_id);
+                let ty = self
+                    .tcx
+                    .instantiate_bound_regions_with_erased(Binder::bind_with_vars(ty, bound_vars));
+                let ty = match self.tcx.asyncness(fn_id.owner) {
+                    ty::Asyncness::Yes => self.get_impl_future_output_ty(ty).unwrap_or_else(|| {
+                        span_bug!(
+                            fn_decl.output.span(),
+                            "failed to get output type of async function"
+                        )
+                    }),
+                    ty::Asyncness::No => ty,
+                };
+                let ty = self.normalize(expr.span, ty);
+                self.can_coerce(found, ty)
+            }
+            hir::FnRetTy::DefaultReturn(_) if in_closure => {
+                let mut rets = vec![];
+                if let Some(ret_coercion) = self.ret_coercion.as_ref() {
+                    let ret_ty = ret_coercion.borrow().expected_ty();
+                    rets.push(ret_ty);
+                }
+                let mut visitor = CollectRetsVisitor { ret_exprs: vec![] };
```

Comment:
Why do you need to do this? If the `ret_coercion` has a type, then it will have already been coerced with all of these `ret_exprs` collected below.

Reply:
Like, the whole point of `ret_coercion` is to be the type that all of the `return` expressions in the body are coerced into

Reply:
`ret_coercion` is (always?) an inference type.

Reply:
It's not always an inference type. It may *appear* to be an inference type if you print it without calling `resolve_vars_if_possible` on it (which you don't need to do here), but it gets equated with another type as soon as the first `return` statement is type checked, and `can_coerce` will handle it properly below.

Applying this diff makes no changes to the UI:
```diff
diff --git a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
index afe5d0d4880..0a03429b37f 100644
--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -1,6 +1,5 @@
 use super::FnCtxt;
 
-use crate::coercion::CollectRetsVisitor;
 use crate::errors;
 use crate::fluent_generated as fluent;
 use crate::fn_ctxt::rustc_span::BytePos;
@@ -17,7 +16,6 @@
 use rustc_hir as hir;
 use rustc_hir::def::Res;
 use rustc_hir::def::{CtorKind, CtorOf, DefKind};
-use rustc_hir::intravisit::{Map, Visitor};
 use rustc_hir::lang_items::LangItem;
 use rustc_hir::{
     CoroutineDesugaring, CoroutineKind, CoroutineSource, Expr, ExprKind, GenericBound, HirId, Node,
@@ -1082,23 +1080,6 @@ pub(in super::super) fn suggest_missing_break_or_return_expr(
                     let ret_ty = ret_coercion.borrow().expected_ty();
                     rets.push(ret_ty);
                 }
-                let mut visitor = CollectRetsVisitor { ret_exprs: vec![] };
-                if let Some(item) = self.tcx.hir().find(id)
-                    && let Node::Expr(expr) = item
-                {
-                    visitor.visit_expr(expr);
-                    for expr in visitor.ret_exprs {
-                        if let Some(ty) = self.typeck_results.borrow().node_type_opt(expr.hir_id) {
-                            rets.push(ty);
-                        }
-                    }
-                    if let hir::ExprKind::Block(hir::Block { expr: Some(expr), .. }, _) = expr.kind
-                    {
-                        if let Some(ty) = self.typeck_results.borrow().node_type_opt(expr.hir_id) {
-                            rets.push(ty);
-                        }
-                    }
-                }
                 rets.into_iter().all(|ty| self.can_coerce(found, ty))
             }
             _ => false,
```

Unless I am mistaken, pls remove the `CollectRetsVisitor` usage (and the changes to it).

Reply:
For the record, that diff is to just demonstrate that it works without the usage of `CollectRetsVisitor `. The code should be simplified into a single `can_coerce` call.

Reply:
Actually, I think none of your tests aren't actually exercising this behavior of a previous return statement having constrained the `ret_coercion`?

Anyways, you should also be using `.merged_ty()` from the `ret_coercion` instead of `.expected_ty()`, but you still shouldn't need to look at any of the `node_ty`s of the return expressions. You should write a test that exercises that the lint will *not* be made if the type is incompatible with other returns.

Reply:
I wanted to avoid an incorrect suggestion in the face of

```rust
pub fn baz() -> Pin<Box<dyn Future<Output = Result<S, ()>> + 'static>> {
    Box::pin(async move {
        if true {
            Some(S) //~ ERROR mismatched types
        }
        Err(())
    })
}
```

Now, looking into it, not only the logic is now wrong (after doing some changes cleaning up for publication), but the type for the returning expressions isn't actually available. For some reason, I *had* gotten a value for them, but now with the code as is now. Given that, yes, the entire visitor thing is useless and can be removed. I was hoping to avoid nonsensical suggestions like the one above, but oh well.

Reply:
Keeping around a single can_coerce call for the `ret_coercion.expected_ty()` should avoid such a nonsensical suggestion, at least according to local testing.

Reply:
I think I know why I had something that seemed to be working earlier. I tried using `check_expr` at some point, which "worked", but trying it again now it does cause multiple issues that make it more of a pain in the neck than we want to deal with.